From 511e2977528151e12c5f49572c79d97ef6289a67 Mon Sep 17 00:00:00 2001
From: Colin Patrick Mccabe <cmccabe@cloudera.com>
Date: Wed, 15 Jun 2016 22:47:01 -0700
Subject: [PATCH 1616/2748] HDFS-10525. Fix NPE in
 CacheReplicationMonitor#rescanCachedBlockMap
 (Xiao Chen via cmccabe)

(cherry picked from commit 2ca73445f5c2929d9c2ff4232dca58a63a0570a0)
(cherry picked from commit 8280e3bcc21adbbbb21e091bd4ec786bf3beb72b)
(cherry picked from commit ea2e7321d6c9b333a8decbee6f7d5ab2d03243b4)

Change-Id: I8d1a3356f727a7a9f60fde2ae70a4fc46d5425ea
---
 .../blockmanagement/CacheReplicationMonitor.java   |    8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/CacheReplicationMonitor.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/CacheReplicationMonitor.java
index a8c4a85..ae3c1db 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/CacheReplicationMonitor.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/CacheReplicationMonitor.java
@@ -504,6 +504,14 @@ private void rescanCachedBlockMap() {
         CachedBlock cblock = it.next();
         BlockInfo blockInfo = blockManager.
             getStoredBlock(new Block(cblock.getBlockId()));
+        if (blockInfo == null) {
+          // Cannot find this block on the NameNode, skip this block from
+          // capacity calculation. Later logic will handle this block.
+          LOG.debug("Block {}: cannot be found in block manager and hence"
+              + " skipped from calculation for node {}.", cblock.getBlockId(),
+              dn.getDatanodeUuid());
+          continue;
+        }
         if (blockInfo.getNumBytes() > remaining) {
           LOG.debug("Block {}: removing from PENDING_CACHED for node {} "
                   + "because it cannot fit in remaining cache size {}.",
-- 
1.7.9.5

